### PR TITLE
Speed up tests

### PR DESCRIPTION
**Pull request by vineetbansal**
The tests are currently taking 1hr53mins on my local machine (even with a gpu). I think this will be a detriment to development going forward. This PR will try to reduce the testing time to ~5 minutes or less through a series of commits.

### CODE REVIEW

File: src/paste3/helper.py [hunk 1]
```diff
@@ -214,3 +233,29 @@ def kl_divergence_backend(X, Y):
     X_log_X = nx.reshape(X_log_X, (1, X_log_X.shape[0]))
     D = X_log_X.T - nx.dot(X, log_Y.T)
     return nx.to_numpy(D)
+
+
+def dissimilarity_metric(which, sliceA, sliceB, A, B, **kwargs):
```
Comment (vineetbansal) [inline src/paste3/helper.py:238]: Since dissimilarity calculation is a major time-consuming component of partial_align, I've moved this to a separate function so it can be tested independently for all algorithms, and we need only test the partial_align function for a single algorithm (with mocked values for this function when its encountered)

File: src/paste3/model_selection.py [hunk 2]
```diff
@@ -144,95 +144,6 @@ def calculate_convex_hull_edge_inconsistency(sliceA, sliceB, pi):
     return measure_A, measure_B
 
 
-"""
-Main function.
-"""
-
-
-def select_overlap_fraction(sliceA, sliceB, alpha=0.1):
```
Comment (vineetbansal) [inline src/paste3/model_selection.py:152]: `select_overlap_fraction_plotting` was simply a copy of this function with one additional parameter (`show_plot`) which only adds to the functionality as an added side-effect, and doesn't modify the rest of the function's behavior. So that has been renamed to `select_overlap_fraction`, and this function is gone. The associated test for the function is gone too.

File: src/paste3/paste.py [hunk 3]
```diff
@@ -146,9 +140,13 @@ def pairwise_align(
     else:
         b = nx.from_numpy(b_distribution)
 
-    if isinstance(nx, ot.backend.TorchBackend) and use_gpu:
-        a = a.cuda()
-        b = b.cuda()
+    if isinstance(nx, ot.backend.TorchBackend):
```
Comment (vineetbansal) [inline src/paste3/paste.py:143]: Grouped all conditionals for `TorchBackend` in one place. The `a = a.double()` is needed because of the difference between the returned dtypes between `ot.backend.NumpyBackend().ones(..)` (double-precision)) and `ot.backend.TorchBackend.ones(..)` (single-precision)

File: src/paste3/paste.py [hunk 4]
```diff
@@ -104,9 +104,6 @@ def pairwise_align(
     coordinatesB = sliceB.obsm["spatial"].copy()
     coordinatesB = nx.from_numpy(coordinatesB)
 
-    if isinstance(nx, ot.backend.TorchBackend):
```
Comment (vineetbansal) [inline src/paste3/paste.py:107]: I suspect this code was put in ad-hoc because the math operations weren't being called correctly because of dtypes being mixed when using `TorchBackend` (numpy is more forgiving of mixing dtypes than torch).

This code was the first indication why tests were failing on macs (reducing precision this way will cause platform-specific differences to magnify). Thus these kinds of `.float()` conversions (at 3 places) have been jettisoned.

File: src/paste3/model_selection.py [hunk 5]
```diff
@@ -252,7 +163,7 @@ def plot_edge_curve(m_list, source_list, target_list):
     plt.show()
 
 
-def select_overlap_fraction_plotting(sliceA, sliceB, alpha=0.1, show_plot=True):
+def select_overlap_fraction(sliceA, sliceB, alpha=0.1, show_plot=True, numItermax=1000):
```
Comment (vineetbansal) [inline src/paste3/model_selection.py:166]: Here and at several other places, `numItermax` (or `maxIter`) have been exposed (while keeping the original default values) so we can override them in tests.

File: tests/conftest.py [hunk 6]
```diff
@@ -1,30 +1,15 @@
...
 import pytest
 from paste3.helper import intersect
-import torch
 import ot.backend
 
 test_dir = Path(__file__).parent
 input_dir = test_dir / "data/input"
 
 
-def pytest_generate_tests(metafunc):
```
Comment (vineetbansal) [inline tests/conftest.py:13]: After we modified test to check on fewer combination of parameters, the remaining parametrization only happens on a single combination of parameters, so we don't need to parametrize anyway.

File: tests/conftest.py [hunk 7]
```diff
@@ -34,7 +19,7 @@ def slices():
         c_fpath = Path(f"{input_dir}/slice{i}_coor.csv")
 
         # Create ann data object of each slice and add other properties
-        _slice = sc.read_csv(s_fpath)
+        _slice = sc.read_csv(s_fpath, dtype="float64")
```
Comment (vineetbansal) [inline tests/conftest.py:22]: needed for correct dtype operations in the code (NumpyBackend hid away this discrepancy for us initially).

File: tests/conftest.py [hunk 8]
```diff
@@ -84,3 +69,10 @@ def slices2():
         slices.append(_slice)
 
     return slices
+
+
+@pytest.fixture(scope="function", autouse=True)
```
Comment (vineetbansal) [inline tests/conftest.py:74]: I think this is useful in all cases, and avoids `np.seed()` calls in individual tests.

File: tests/test_model_selection.py [hunk 9]
```diff
@@ -52,24 +51,17 @@ def test_edge_inconsistency_score(slices, tmp_path):
 
 
 def test_calculate_convex_hull_edge_inconsistency(slices, tmp_path):
-    np.random.seed(0)
-    pairwise_info = partial_pairwise_align(slices[0], slices[1], s=0.7)
+    pairwise_info = partial_pairwise_align(slices[0], slices[1], s=0.7, maxIter=10)
```
Comment (vineetbansal) [inline tests/test_model_selection.py:54]: In a lot of tests, oracle values have been recomputed by using `maxIter` (or `numItermax`) as 10.

File: tests/test_paste2.py [hunk 10]
```diff
@@ -19,85 +20,28 @@
 output_dir = test_dir / "data/output"
 
 
-def pytest_generate_tests(metafunc):
```
Comment (vineetbansal) [inline tests/test_paste2.py:22]: I feel that metafunctions like this should only be used if the same parametrization is used in many places (let's say 3 or more). Otherwise its better to just put the decorator right where its used, reducing the need to navigate somewhere else in the codebase to see what's going on.

File: tests/test_paste2.py [hunk 11]
```diff
@@ -19,85 +20,28 @@
...
-                        "partial_fgw_cost": 30.76272781211168,
-                    },
-                    "partial_fused_gromov_wasserstein_true.csv",
-                ),
-            ],
-        )
-
-
-def test_partial_pairwise_align(slices2, dissimilarity, filename):
+@patch("paste3.paste2.dissimilarity_metric")
```
Comment (vineetbansal) [inline tests/test_paste2.py:23]: Since `dissimilarity_metric` is being tested elsewhere, its return value can be mocked here.

File: tests/test_paste_helpers.py [hunk 12]
```diff
@@ -215,3 +215,35 @@ def test_norm_and_center_coordinates(slices):
         check_dtype=False,
         rtol=1e-04,
     )
+
+
+@pytest.mark.parametrize(
```
Comment (vineetbansal) [inline tests/test_paste_helpers.py:220]: New test purely to check the newly introduced dissimilarity_metric function

**Review by anushka255 — APPROVED** [review]
